Add `#gallery` to `Variant` and `Product`

swcraig opened the pull request:
[Paperclip is deprecated](https://robots.thoughtbot.com/closing-the-trombone) and we need to remove it from stock Solidus.

In order to remove Paperclip properly we need to add extension points so that we can plug a different image uploading library (like Active Storage). Adding galleries to `Product` and `Variant` will give us more control over what we send to the view layer, and protect our core models from image implementation details.

To plug this interface a developer needs to create custom galleries for both `Variant` and `Product` and make sure that the `#images` method of those return a collection of images that respond to the new "presentational" image interface the frontend expects. This PR includes changes to `Spree::Image` which transitions it into a presentation role for Paperclip. Specifically, the presentational image interface that the view layer expects is:

```
id: int
url(:size): string
filename: string
viewable_id: int # the ID of the variant this is "attached" to
```

Before Paperclip can be extracted into its own extension properly we need this PR to be merged.

Much of the work around the the `Gallery` classes are heavily influenced by/verbatim the work Clarke did in #625.

jhawthorn commented on core/app/models/spree/variant.rb:
```diff
@@ -350,6 +350,16 @@ def variant_properties
+    # An Enumerable of media that goes along with a variant
+    # Stores may not necessarily want to attach images to variants
+    # This provides a more robust interface
+    # NOTE: In the future this will return an empty array
```
Why would this be useful? What is a developer supposed to do with this information.

I assume this is for extension points, since we want to avoid `class_eval`ing, maybe we should look at introducing a `Spree::Config.variant_gallery_class` and a `Spree::Config.product_gallery_class`.

tvdeyen replied:
I like that idea of configurable image attachment classes. If we want extensions to be able to hook into we want to provide a real integration point. 

Also I don’t think `gallery` is the correct term. `gallery_images` would follow the “plural is a collection” convention of Rails better. We could even override the `images` getter. Btw. We need to deprecate the `images =` setter as well or delegate this to the images provider class.

jhawthorn commented on core/app/models/spree/gallery/variant_gallery.rb:
```diff
@@ -0,0 +1,25 @@
+      private
+
+      def product_images
+        @variant.product.try(:variant_images) || []
```
I don't think this needs to be a `try`. Is `@variant` ever `nil`?

jhawthorn commented on core/app/models/spree/gallery/base.rb:
```diff
@@ -0,0 +1,23 @@
+module Spree
```
I'm not sure there's a need for this base. It should be sufficient to duck-type have the two gallery classes "quack" similarly :duck: :duck: :duck: :duck: :duck:

swcraig replied:
:duck: 

I guess the reason I would want the base class is so that it's obvious the interface that should be implemented for galleries in general. But I understand your point and can make the changes.

swcraig commented:
This is ready for another review. There have been a few changes since the last reviews. I have added `Variant#gallery.best_image`. This method will fallback to using the first associated product image if a variant itself has no images.

tvdeyen commented on core/app/models/spree/gallery/variant_gallery.rb:
```diff
@@ -0,0 +1,38 @@
+      private
+
+      def product_images
+        @variant.product.variant_images
```
I don’t see a need for this private method as long as we don’t access it from two different methods at least.

tvdeyen commented on core/app/models/spree/gallery/variant_gallery.rb:
```diff
@@ -0,0 +1,38 @@
+      # @return [Spree::Image] best image for the variant gallery
+      def best_image
+        @variant.images.first || product_images.first || Spree::Image.new
+      end
```
Why not remove these image collector methods from here and move into the views. This might be more duplication, but is also more flexible. Although I think the naming is difficult and misleading. Let’s not make these decisions here, but in the views. They are easier to override in a per use case basis.

swcraig commented:
I think this is good to go. I've also added `ImagesHelper#image_or_default` for the admin/frontend views.

swcraig commented:
Thank you for the feedback.
 
The galleries have an `#image_urls` method in addition to an the `#images` method. The JBuilder API endpoints. the product thumbnails partial on the frontend, and `images/_image_row`  in the backend expect `Spree::Image`s (that respond to Paperclip API calls). I can't see an easy way to rewrite these with only URLs and keep the same behaviour.

The actual display logic no longer needs to know Paperclip's API (there is no more accessing of the `#attachment` of an image). Another sort of unfortunate result is that the views need to specify the image size twice (in the case that there is no image and we need to fallback to the CSS of `class= placeholder <size>`. [Like here for example](https://github.com/swcraig/solidus/blob/20d4b5cad2f5a7eedc9285c69c9639263f2b6226/backend/app/views/spree/admin/cancellations/index.html.erb#L28-L30).

I think this is closer to getting rid of the Paperclip specific logic out of the views, and I'm interested in hearing what we think of where this is going.

mamhoff commented on backend/app/views/spree/admin/shared/_image.html.erb:
```diff
@@ -1,7 +1,7 @@
-<% if image && image.attachment? %>
-  <%= image_tag image.attachment(size) %>
+<% if image %>
+  <%= image_tag image %>
```
This is also missing the `:mini` option.

swcraig replied:
This partial is given an image url (the caller has already decided the size and retrieved the relevant url). I've updated the naming to be clearer.

Does that make sense?

tvdeyen commented on core/app/models/spree/gallery/product_gallery.rb:
```diff
@@ -0,0 +1,16 @@
+      #
+      # @return [Enumerable<Spree::Image>] all images in the gallery
+      def images
+        @images ||= (@product.images + @product.variant_images).uniq
```
There is no need to merge these two collections, because `@product.variant_images` already [includes images of the master variant](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/product.rb#L75) - what are [the images returned by `@product.images`](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/product.rb#L72)

tvdeyen commented on backend/app/views/spree/admin/stock_items/_stock_management.html.erb:
```diff
@@ -28,7 +28,9 @@
             <div class='variant-image'>
-              <%= render 'spree/admin/shared/image', image: variant.display_image(fallback: false), size: :small %>
+              <%= render 'spree/admin/shared/image',
+                image: variant.gallery.images.first.try(:url, :small),
```
The thing you're passing into the partial as `image` is actually an `url`. Shouldn't we rename the variable as well?

tvdeyen commented on backend/app/views/spree/admin/orders/_shipment_manifest.html.erb:
```diff
@@ -4,7 +4,9 @@
     <td class="item-image align-center">
-      <%= render 'spree/admin/shared/image', image: item.variant.display_image, size: :mini %>
+      <%= render 'spree/admin/shared/image',
+        image: (item.variant.gallery.images.first || item.variant.product.gallery.images.first).try(:url, :mini),
```
It makes sense to have a `first_image_url` method in the gallery classes.

```rb
item.variant.gallery.first_image_url(size: :mini) || item.variant.product.gallery.first_image_url(size: :mini)
```

tvdeyen commented on core/app/models/spree/gallery/product_gallery.rb:
```diff
@@ -0,0 +1,16 @@
+      # A list of all images associated with this gallery
+      #
+      # @return [Enumerable<Spree::Image>] all images in the gallery
+      def images
```
I think we would gain much more isolation from  this class if it would have an `image_urls` method that takes a `size:` parameter.

tvdeyen commented on core/app/models/spree/gallery/variant_gallery.rb:
```diff
@@ -0,0 +1,16 @@
+      # A list of all images associated with this gallery
+      #
+      # @return [Enumerable<Spree::Image>] all images in the gallery
+      def images
```
Here as well, a `image_urls` method that takes a `size:` parameter would help us decoupling from paperclip

tvdeyen commented on core/app/models/spree/image.rb:
```diff
@@ -19,9 +19,18 @@ class Image < Asset
+    end
+
+    def url(size)
+      attachment.url(size)
```
We are not able to decouple from paperclip if we have this method inside of here. It would be better to put this url retrieving logic inside the gallery classes. Extensions like `solidus_paperclip` would then implement such gallery classes.

swcraig commented on frontend/app/views/spree/products/_thumbnails.html.erb:
```diff
@@ -1,17 +1,19 @@
-        <%= link_to(image_tag(i.attachment.url(:mini)), i.attachment.url(:product)) %>
+
+    <% @product.gallery.images.each do |image| %>
+      <% next if image.viewable_id != @product.master.id %>
```
I don't see how I am going to be able to write this `_thumbnails` partial using only image URLs. At some point an image is going to need the context of which variant it is for. What do you think, @tvdeyen?

If we have situations like this (and the `admin/_image_row` partial), I think it would make sense for the galleries to return an "Image" class. If a developer wants to drop in a different implementation they can provide their own image class that responds to `#url`, `#id`, `#viewable_id`, and `#alt`. I was thinking something like this:

```
class MyImageClass
  def initialize(image)
    @image = image
  end

  def url(size)
    @image.the_way_this_image_library_gets_urls(size)
  end

  # etc...
end

class MyVariantGallery
  def images
    however_the_images_are_stored.map { |image| MyImage.new(image) }
  end
end
```

tvdeyen replied:
I like the idea of a separate `Image` class 👌

So we would have a repository (`Gallery`) and a presenter (maybe `Picture`?). This sounds like a good plan to me.

swcraig replied:
After thinking about this, I think that we can transition `Spree::Image` to function as this presenter. I don't think there is much value in adding another class (`Spree::Picture`), when we can use `Spree::Image` to do the same thing. I've updated the PR to reflect this.

swcraig commented on frontend/app/views/spree/shared/_image.html.erb:
```diff
@@ -1,11 +1,11 @@
+<% if image_url %>
   <% if itemprop %>
-    <%= image_tag image.attachment(size), itemprop: itemprop %>
+    <%= image_tag image_url, itemprop: itemprop %>
```
Something to note here. We don't use an image's `alt` attribute in either frontend or backend. Would being able to access `image.alt` here be an argument for using a custom image class [like I suggest in this PR comment](https://github.com/solidusio/solidus/pull/2337/files#r154390915)? That image class could have a strict interface `#url(size)`, `#id` (for that single-use modal on `backend/image/_image_row`), `#viewable_id`, and `#alt`.

Is there a reason why we haven't historically been rendering an image's `alt` attribute? @mamhoff @tvdeyen

tvdeyen replied:
> Would being able to access image.alt here be an argument for using a custom image class

Yes! A separate image class (`Picture` maybe?) makes total sense to abstract the presenter API 👍

tvdeyen replied:
> Is there a reason why we haven't historically been rendering an image's alt attribute?

Not sure. We should render it (at least in the frontend).

swcraig commented:
I have this to the point where we don't reference Paperclip specific calls anywhere in the views.  We do continue to have Paperclip specific logic in the [solidus_api controller for images](https://github.com/swcraig/solidus/blob/147c57aac515f072fd09734dd4954e8688fdb518/api/app/controllers/spree/api/images_controller.rb), and in the [actual form for uploading images](https://github.com/swcraig/solidus/blob/147c57aac515f072fd09734dd4954e8688fdb518/backend/app/views/spree/admin/images/_form.html.erb). We can probably deal with these when someone goes to pull Paperclip into an extension.

swcraig commented:
This is ready for another review. I plan on [transitioning `Spree::Image` to function as a "presenter" class for images on the frontend](https://github.com/solidusio/solidus/pull/2337#discussion_r156772037). It's sort of weird right now because it represents a Paperclip image, but instead of getting rid of it and using a new model (`Spree::Picture`) if/when someone rips out Paperclip we could just keep it and make it define the interface that views understand.

swcraig commented:
Could we possibly take another look at this? With [Rails 5.2/ActiveStorage close to release](http://weblog.rubyonrails.org/2018/1/30/Rails-5-2-RC1-Active-Storage-Redis-Cache-Store-HTTP2-Early-Hints-Credentials/) I think there should be some consideration to get Paperclip out.

swcraig commented:
[With Paperclip officially being deprecated](https://robots.thoughtbot.com/closing-the-trombone) can we take a serious look at merging this? When this is merged it will make extracting Paperclip into an extension much easier. I also think this is a better general approach to dealing with images in core/the view layer.

@tvdeyen feedback would be appreciated.

tvdeyen commented:
🎉 🎁
